### ISSUE TITLE
Shared JIT classloader does not account for duplicate class names

### ISSUE THREAD

**Issue by headius** [jruby/jruby#6281]
When enabling the "shared" mode for the `jit.loader.mode` property, intended to load all JIT bytecode into the same classloader, I see the following `jit.logging` output for a script with two methods:

```
$ jruby -Xjit.loader.mode=SHARED -Xjit.logging foo_bar_inline.rb
2020-06-14T06:48:17.452-05:00 [Ruby-0-JIT-1] INFO JITCompiler : method done jitting: Object foo at foo_bar_inline.rb:6
2020-06-14T06:48:17.460-05:00 [Ruby-0-JIT-1] INFO JITCompiler : could not compile method; passes run: [org.jruby.ir.passes.LocalOptimizationPass@584ed071, org.jruby.ir.passes.OptimizeDynScopesPass@f61f6f63, org.jruby.ir.passes.DeadCodeElimination@805d30ee, org.jruby.ir.passes.OptimizeDelegationPass@6c7a3ce9, org.jruby.ir.passes.AddCallProtocolInstructions@87f2d916, org.jruby.ir.passes.AddMissingInitsPass@693f2559]: Bar bar at foo_bar_inline.rb:1 because of: "java.lang.LinkageError: loader (instance of  org/jruby/util/ClassDefiningJRubyClassLoader): attempted  duplicate class definition for name: "foo_bar_inline""
```

Note that this run of JRuby has been modified to not immediately compile the target script, while investigating #6280. The script combination there may show the same error.

I believe this is due to the name of the script being used for the class name of the jitted method, which then conflicts if more than one method gets jitted separately but loaded into the same classloader.

We will want to use a different name-mangling for jit classes when we know they'll be loaded into the same classloader.

Ping @kares for thoughts since he added this option.

**Comment by kares** [comment]
right, makes sense - have assumed JIT-ed class names to be unique within the same runtime.

**Comment by headius** [comment]
@kares There's no reason they couldn't be... they just aren't. Never noticed it before because we always used isolated classloaders.

I did a five minute patch to add a monotonic index to the name and it works fine. We just need to do it in a more formal way.
